### PR TITLE
[TS migration] Migrate 'EnablePayments' pages to TypeScript

### PR DESCRIPTION
**Pull request by jayeshmangwani**
### Fixed Issues

$ https://github.com/Expensify/App/issues/25222
PROPOSAL: https://github.com/Expensify/App/issues/25222#issuecomment-1912037055

### Tests

1. Go to Settiings > Wallet > Enable Wallet > Add a bank account
2. Now we will be navigated to Additional details page, verify the working of the Additional details page with existing functionality (e.g. form should show error on empty fields and pressing button should navigate to next page)
3. To verify the IdologyQuestions page, paste below code to console on the on the Additional details step, and verify that IdologyQuestions step is working same as the current staging version 
```javascript
Onyx.merge('walletAdditionalDetails', {questions: [{"answer":["1251","6253","113","None of the above","Skip Question"],"prompt":"Which number goes with your address on MASONIC AVE?","type":"street.number.b"}]})
```
4. Next page will be OnfidoStep page, and verify that page loads perfectly and verify that OnfidoStep is working same as existing app.

- [x] Verify that no errors appear in the JS console

### Offline tests

Enable Payment Flow is disabled in offline

### QA Steps

1. Go to Settiings > Wallet > Enable Wallet > Add a bank account
2. Add details on Additional details page, verify the working of the Additional details page with current staging version.
3. Verify the Onfido page, Idology questions step is working same as the current staging version.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/35371050/62269231-d096-4937-8814-bbbf0cda6bb9

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/35371050/21dfb0fa-41ac-489d-9ad6-56a687f41e51

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/35371050/d65bbcc9-4202-42fd-a3e0-b0d955e36317

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/35371050/be5f0844-897d-49ac-ac5f-42ad74492f49

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/35371050/b12fcdc8-3767-435f-afab-ea98fc60550c

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/35371050/1541587a-315d-412f-8d49-4a8379a4ef50

</details>

### CODE REVIEW

**Comment by jayeshmangwani** [comment]
@ntdiary code has been migrated, and I am working on the testing the individual pages, I will update the Tests section and the attachments once I can run the app on my side, getting some weird issue `Unable to resolve reference $expo`

**Comment by jayeshmangwani** [comment]
@ntdiary I have added the additional details to enable-payments page and navigated to OnfidoStep, so not sure how to bypass the onfido verification, do you have idea how can we test all the pages under enable-payments page after adding the additional details page?

https://github.com/Expensify/App/assets/35371050/68ecb2f9-f17c-49dd-a6d5-46d1527f39ba

File: src/ONYXKEYS.ts [hunk 1]
```diff
@@ -557,6 +561,10 @@ type OnyxValues = {
     [ONYXKEYS.FORMS.REIMBURSEMENT_ACCOUNT_FORM_DRAFT]: OnyxTypes.Form;
     [ONYXKEYS.FORMS.PERSONAL_BANK_ACCOUNT]: OnyxTypes.PersonalBankAccount;
     [ONYXKEYS.FORMS.PERSONAL_BANK_ACCOUNT_DRAFT]: OnyxTypes.PersonalBankAccount;
+    [ONYXKEYS.FORMS.IDOLOGY_QUESTIONS_FORM]: OnyxTypes.IdologyQuestionsForm;
+    [ONYXKEYS.FORMS.IDOLOGY_QUESTIONS_FORM_DRAFT]: OnyxTypes.IdologyQuestionsForm;
+    [ONYXKEYS.FORMS.WALLET_ADDITIONAL_DETAILS]: OnyxTypes.AdditionalDetailStepForm;
```
Comment (blazejkustra) [inline src/ONYXKEYS.ts:566]: Good idea with adding new Form keys 👍

File: src/ONYXKEYS.ts [hunk 2]
```diff
@@ -367,6 +367,10 @@ const ONYXKEYS = {
         REIMBURSEMENT_ACCOUNT_FORM_DRAFT: 'reimbursementAccountDraft',
         PERSONAL_BANK_ACCOUNT: 'personalBankAccountForm',
         PERSONAL_BANK_ACCOUNT_DRAFT: 'personalBankAccountFormDraft',
+        IDOLOGY_QUESTIONS_FORM: 'idologyQuestions',
```
Comment (blazejkustra) [inline src/ONYXKEYS.ts:370]: How about:
```suggestion
        IDOLOGY_QUESTIONS_FORM: 'idologyQuestionsForm',
```

**Review by blazejkustra — APPROVED** [review]
Great work typescript-wise 💯

File: src/ONYXKEYS.ts (hunk 2 above)
Comment (jayeshmangwani) [reply src/ONYXKEYS.ts]: Yes it's a good change, now updated

**Comment by jayeshmangwani** [comment]
@ntdiary merge conflicts resolved and PR author checklist completed , please help in review

**Review by ntdiary — CHANGES_REQUESTED** [review]

File: src/ONYXKEYS.ts [hunk 3]
```diff
@@ -388,6 +388,10 @@ const ONYXKEYS = {
         REIMBURSEMENT_ACCOUNT_FORM_DRAFT: 'reimbursementAccountDraft',
         PERSONAL_BANK_ACCOUNT: 'personalBankAccountForm',
         PERSONAL_BANK_ACCOUNT_DRAFT: 'personalBankAccountFormDraft',
+        IDOLOGY_QUESTIONS_FORM: 'idologyQuestionsForm',
+        IDOLOGY_QUESTIONS_FORM_DRAFT: 'idologyQuestionsFormDraft',
+        WALLET_ADDITIONAL_DETAILS: 'walletAdditionalDetailsForm',
+        WALLET_ADDITIONAL_DETAILS_DRAFT: 'walletAdditionalDetailsFormDraft',
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/ONYXKEYS.ts:394]: Ideally I would change the keys. If it's not possible, suppressing it seems like acceptable alternative

[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/ONYXKEYS.ts:394]: @ntdiary I am fixing merge conflicts and will revert the WALLET_ADDITIONAL_DETAILS key name today, sorry somehow I missed the notification

[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/ONYXKEYS.ts]: Reverted the WALLET_ADDITIONAL_DETAILS key and pushed the code

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/types/onyx/index.ts [hunk 4]
```diff
@@ -153,5 +154,10 @@ export type {
     RecentlyUsedReportFields,
     LastPaymentMethod,
     InvitedEmailsToAccountIDs,
+    PrivateNotesForm,
+    ReportFieldEditForm,
+    IdologyQuestionsForm,
+    AdditionalDetailStepForm,
+    RoomNameForm,
```
[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/types/onyx/index.ts]: Thanks @blazejkustra , I have removed the old usage

File: src/pages/EnablePayments/IdologyQuestions.tsx [hunk 5]
```diff
@@ -126,13 +106,13 @@ function IdologyQuestions({questions, idNumber}) {
...
                     style={styles.mb3}
-                    href="https://use.expensify.com/usa-patriot-act"
+                    href={CONST.HELP_LINK_URL}
                 >
                     {translate('additionalDetailsStep.helpLink')}
                 </TextLink>
             </View>
             <FormProvider
-                formID={ONYXKEYS.WALLET_ADDITIONAL_DETAILS}
+                formID={ONYXKEYS.FORMS.IDOLOGY_QUESTIONS_FORM}
```
Comment (jayeshmangwani) [inline src/pages/EnablePayments/IdologyQuestions.tsx:115]: We had used the `ONYXKEYS.WALLET_ADDITIONAL_DETAILS` key for IdologyQuestions Form,
https://github.com/Expensify/App/blob/f50ac2af5194c9b3c8ba538180742972c8a41700/src/pages/EnablePayments/IdologyQuestions.js#L134-L135

But in this PR I have created a new key `ONYXKEYS.FORMS.IDOLOGY_QUESTIONS_FORM` for the  IdologyQuestions, but now I thinking that server erorr will not be visible for this Form and I think we have to use the previous(WALLET_ADDITIONAL_DETAILS) key here, @ntdiary you have any other thought on this change ?

cc: @blazejkustra

Comment (blazejkustra) [reply src/pages/EnablePayments/IdologyQuestions.tsx:115]: I think you are right and we have to use the same key. Let's add `ONYXKEYS. FORMS.WALLET_ADDITIONAL_DETAILS` with the same value for the key as `ONYXKEYS.WALLET_ADDITIONAL_DETAILS` has.

You will have to use  `// @ts-expect-error` in some places but that's fine as there will be conflict in these types:`ONYXKEYS.WALLET_ADDITIONAL_DETAILS` vs `ONYXKEYS. FORMS.WALLET_ADDITIONAL_DETAILS`

Comment (jayeshmangwani) [reply src/pages/EnablePayments/IdologyQuestions.tsx]: @blazejkustra Removed IdologyQuestionsForm and now we are using the current AdditionalDetailStepForm and We do not have to add // @ts-expect-error anywhere, as `IdologyQuestions` page does not use the any Onyx key, @ntdiary we have added above the mentioned changes, PR is ready for the review

**Review by ntdiary — CHANGES_REQUESTED** [review]
@jayeshmangwani, just several tiny request changes. :)

File: src/pages/EnablePayments/TermsPage/LongTermsForm.tsx [hunk 6]
```diff
@@ -105,7 +104,6 @@ function LongTermsForm() {
             <View style={styles.flexRow}>
                 <Icon
                     fill={theme.icon}
-                    style={styles.flex1}
```
[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/pages/EnablePayments/TermsPage/LongTermsForm.tsx:108]: this change is intentional, we does not have any prop with name `style` in the Icon and we are not using the prop too, so this line has no effect IMO
https://github.com/Expensify/App/blob/5199604ac28f9db9fdcdc8bf7e8b4ce4e406ded8/src/components/Icon/index.tsx#L12-L45

File: src/pages/EnablePayments/AdditionalDetailsStep.tsx [hunk 7]
```diff
@@ -77,22 +51,27 @@ const fieldNameTranslationKeys = {
...
     ssnFull9: 'common.ssnFull9',
-};
-
-function AdditionalDetailsStep({walletAdditionalDetails, translate, currentUserPersonalDetails}) {
+} as const;
+const STEP_FIELDS = [
+    INPUT_IDS.LEGAL_FIRST_NAME,
+    INPUT_IDS.LEGAL_LAST_NAME,
+    INPUT_IDS.ADDRESS_STREET,
+    INPUT_IDS.ADDRESS_CITY,
```
[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/pages/EnablePayments/AdditionalDetailsStep.tsx:59]: hey thanks for catching this I've missed to add here, added now

File: src/pages/EnablePayments/AdditionalDetailsStep.tsx [hunk 8]
```diff
@@ -156,11 +132,11 @@ function AdditionalDetailsStep({walletAdditionalDetails, translate, currentUserP
             >
                 <HeaderWithBackButton
                     title={translate('additionalDetailsStep.headerTitle')}
-                    onBackButtonPress={() => Wallet.setAdditionalDetailsQuestions(null)}
+                    onBackButtonPress={() => Wallet.setAdditionalDetailsQuestions([], walletAdditionalDetails?.idNumber ?? '')}
```
[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/pages/EnablePayments/AdditionalDetailsStep.tsx]: Done ✅

File: src/libs/actions/Wallet.ts [hunk 9]
```diff
@@ -75,11 +75,12 @@ function setKYCWallSource(source?: ValueOf<typeof CONST.KYC_WALL_SOURCE>, chatRe
 /**
  * Validates a user's provided details against a series of checks
  */
+
```
[the remark this answers is not in the document]
Comment (jayeshmangwani) [reply src/libs/actions/Wallet.ts]: Done ✅

**Comment by jayeshmangwani** [comment]
@ntdiary Thanks for the review, and now suggested changes has been addressed

**Review by ntdiary — APPROVED** [review]
LGTM. :)
BTW, @jayeshmangwani, it seems we can delete `walletTermsPropTypes.js`?

**Comment by jayeshmangwani** [comment]
Yes we can delete the walletTermsPropTypes, done ✅

**Comment by jayeshmangwani** [comment]
@ntdiary merged latest main and it seems conflicts is solved now

**Comment by jayeshmangwani** [comment]
> merged latest main and it seems conflicts is solved now

@ntdiary bump

**Comment by jayeshmangwani** [comment]
> @jayeshmangwani, I've filled the c+ checklist, the next step is internal engineer review. 😂 BTW, the lint test is failed, I think you can add a comment like this:
> 

@ntdiary Fixed the lint issue, Internal engineer was not assigned automatically to PR, thats why I pinged you , I thought if anything was missing or 🙂

**Comment by jayeshmangwani** [comment]
@Gonals Whenever you get time please help with review

**Comment by Gonals** [comment]
On it. Not sure why I didn't get assigned to it

**Review by Gonals — APPROVED** [review]

**Comment by kbecciv** [comment]
@jayeshmangwani The same result - no Ideology questions are presented if use above information

https://github.com/Expensify/App/assets/93399543/923e258e-fe4d-4763-9db8-a3a1eb25edcb

**Comment by stitesExpensify** [comment]
I think it's pretty likely that this deploy blocker is related to this PR since it's the only one that touched the bank account flow in this deploy https://github.com/Expensify/App/issues/38701

**Pull request merged**